subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/BeanNode.java

```diff
@@ -16,35 +16,108 @@
 
 package org.gradle.api.internal.tasks.properties;
 
-import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
-import com.google.common.collect.Iterators;
+import org.gradle.api.Named;
 
 import javax.annotation.Nullable;
-import java.util.Iterator;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
 
-public class BeanNode extends AbstractPropertyNode<BeanNode> {
-    private final Object bean;
+public abstract class BeanNode extends AbstractPropertyNode<BeanNode> {
 
-    public BeanNode(@Nullable String propertyName, Object bean) {
+    public static BeanNode create(@Nullable String propertyName, Object bean) {
+        if (bean instanceof Map<?, ?>) {
+            return new MapBeanNode(propertyName, (Map<?, ?>) bean);
+        }
+        if (bean instanceof Iterable<?>) {
+            return new IterableBeanNode(propertyName, (Iterable<?>) bean);
+        }
+        return new SimpleBeanNode(propertyName, bean);
+    }
+
+    protected BeanNode(@Nullable String propertyName, Class<?> beanClass) {
+        super(propertyName, beanClass);
+    }
+
+    public abstract Object getBean();
+}
+
+abstract class BaseBeanNode<T> extends BeanNode {
+
+    private final T bean;
+
+    protected BaseBeanNode(@Nullable String propertyName, T bean) {
         super(propertyName, Preconditions.checkNotNull(bean, "Null is not allowed as nested property '" + propertyName + "'").getClass());
         this.bean = bean;
     }
 
-    public Object getBean() {
+    @Override
+    public T getBean() {
         return bean;
     }
 
+    protected BeanNode createChildNode(String propertyName, @Nullable Object input) {
+        String qualifiedPropertyName = getQualifiedPropertyName(propertyName);
+        Object bean = Preconditions.checkNotNull(input, "Null is not allowed as nested property '" + qualifiedPropertyName + "'");
```

Comment:
Be careful: now you are creating the message string every time this check runs. We should manually check `== null` instead.

Reply:
I fixed this by using the method `Preconditions.checkNotNull` which taskes error message parameters.